src/coreclr/jit/emitarm64.h

```diff
@@ -325,96 +325,62 @@ union condFlagsImm {
-static code_t insEncodeReg_R_9_to_5(regNumber reg);
-
-// Return an encoding for the specified 'R' register used in '4' thru '0' position.
-static code_t insEncodeReg_R_4_to_0(regNumber reg);
-
-// Return an encoding for the specified 'P' register used in '19' thru '16' position.
-static code_t insEncodeReg_P_19_to_16(regNumber reg);
-
-// Return an encoding for the specified 'P' register used in '3' thru '0' position.
-static code_t insEncodeReg_P_3_to_0(regNumber reg);
+// Returns an encoding for the specified 'V' register used in 'hi' thru 'lo' position.
+template <const size_t hi, const size_t lo>
+static code_t insEncodeReg_V(regNumber reg)
+{
+    // lo <= hi < 32
+    static_assert((hi >= lo) && (hi < sizeof(code_t) * BITS_PER_BYTE));
+    assert(isVectorRegister(reg));
+    code_t ureg = (code_t)reg - (code_t)REG_V0;
 
-// Return an encoding for the specified 'P' register used in '8' thru '5' position.
-static code_t insEncodeReg_P_8_to_5(regNumber reg);
+    constexpr size_t bits = hi - lo + 1;
+    static_assert(bits <= 5);
+    assert((ureg >= 0) && (ureg < (1 << bits)));
+    return ureg << lo;
+}
 
-// Return an encoding for the specified 'P' register used in '13' thru '10' position.
-static code_t insEncodeReg_P_13_to_10(regNumber reg);
+// Returns an encoding for the specified 'P' register used in 'hi' thru 'lo' position.
+template <const size_t hi, const size_t lo, const bool isHighReg = false>
+static code_t insEncodeReg_P(regNumber reg)
+{
+    // lo <= hi < 32
+    static_assert((hi >= lo) && (hi < sizeof(code_t) * BITS_PER_BYTE));
+    assert(isPredicateRegister(reg));
+    code_t ureg = (code_t)reg - (code_t)REG_P0;
 
-// Return an encoding for the specified 'R' register used in '17' thru '16' position.
-static code_t insEncodeReg_R_17_to_16(regNumber reg);
+    if (isHighReg)
+    {
+        assert(isHighPredicateRegister(reg));
+        ureg -= 8;
+    }
 
-// Return an encoding for the specified 'P' register used in '7' thru '5' position.
-static code_t insEncodeReg_P_7_to_5(regNumber reg);
+    constexpr size_t bits = hi - lo + 1;
+    static_assert(bits <= 4);
+    assert((ureg >= 0) && (ureg < (1 << bits)));
+    return ureg << lo;
+}
 
-// Return an encoding for the specified 'P' register used in '3' thru '1' position.
-static code_t insEncodeReg_P_3_to_1(regNumber reg);
+// Returns an encoding for the specified 'R' register used in 'hi' thru 'lo' position.
+template <const size_t hi, const size_t lo>
+static code_t insEncodeReg_R(regNumber reg)
+{
+    // lo <= hi < 32
+    static_assert((hi >= lo) && (hi < sizeof(code_t) * BITS_PER_BYTE));
+    assert(isIntegerRegister(reg));
+    code_t ureg = (code_t)reg;
 
-// Return an encoding for the specified 'P' register used in '2' thru '0' position.
-static code_t insEncodeReg_P_2_to_0(regNumber reg);
+    constexpr size_t bits = hi - lo + 1;
+    static_assert(bits <= 5);
+    return ureg << lo;
```

Comment:
Should this func have a `assert((ureg >= 0) && (ureg < (1 << bits)));` ?

Reply:
I wanted to include that assert, but some of the methods this is replacing (like `insEncodeReg_R_17_to_16`) have odd invariants around the range of registers they accept (like `>= REG_R12 && <= REG_R15`) that make it awkward to craft a correct assert. We could keep `insEncodeReg_R_17_to_16` around and handle the general case in the templated implementation, or we could add an optional templated arg that specifies the lower bound of the register (so the assertion would check `ureg + lowerBound`, instead of just `ureg`). Any suggestions?

Reply:
Ok - we can't assert because we're only using the low bits of the register number. Happy with that part now.

But, I think we still have an issue (and in existing code). If we pass in a value from 12 to 15 into `insEncodeReg_R_17_to_16` or `insEncodeReg_R<17,16>` then the upper parts of the register number are not being stripped out, and it ends up overflowing into bits 18 to 20. It only works today because in the encode function we then call `insEncodeSveElemsize_tszh_tszl_and_imm` afterwards which rewrites those upper bits.

I think the `insEncodeReg_` functions need to mask `ureg` to the number of bits.

Optionally this could be further expanded with:

```
static code_t insEncodeReg_R(regNumber reg DEBUGARG(bool lowerbitsonly = true))
{
   .....

#if DEBUG
   if (lowerbitsonly)
       assert(ureg && !bitmask == 0)
   else
#endif
       ureg = ureg && bitmask;

 .....
}
```

Reply:
I see, good catch. I'm thinking we can add an optional templated arg for the bitmask, and for previous callers of `insEncodeReg_R_17_to_16`, we can pass a bitmask for extracting the lower two bits. I think `insEncodeReg_P` should use the same approach for encoding high predicate registers.

Reply:
@a74nh I updated the encoder methods to extract only as many bits from `ureg` that will fit within the range specified by `hi` and `lo`. This should keep us from overflowing the register bits in the instruction encoding in the future.

Reply:
With the new version, there are no register range asserts in the encode function, instead it just clips the value to the right size. Meanwhile the sanitycheck function will assert the register range. That makes sense because the sanitycheck is based on the group whereas the encoding function only knows about bits.

Ok, I'm happy with this.